processor/processor_event_dropping_test.go

```diff
@@ -0,0 +1,549 @@
+package processor_test
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	kithttputil "github.com/rudderlabs/rudder-go-kit/httputil"
+	kithelper "github.com/rudderlabs/rudder-go-kit/testhelper"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource/postgres"
+	"github.com/rudderlabs/rudder-server/runner"
+	"github.com/rudderlabs/rudder-server/testhelper/backendconfigtest"
+	"github.com/rudderlabs/rudder-server/testhelper/health"
+	"github.com/rudderlabs/rudder-server/testhelper/transformertest"
+)
+
+func TestProcessorEventDropping(t *testing.T) {
+	t.Run("bot events with drop action should be dropped", func(t *testing.T) {
+		new(eventDropScenario).
+			withEventConfigs([]eventConfig{
+				newBotEventConfig("track", "TestEvent", "drop"),
+				newBotEventConfig("identify", "", "drop"),
+				newBotEventConfig("screen", "", "drop"),
+			}).
+			run(t, func(t *testing.T, scenario *eventDropScenario) {
+				scenario.requireJobsCount(t, "gw", "succeeded", 3)
+				scenario.requireJobsCount(t, "rt", "succeeded", 0)
+				scenario.requireJobsCount(t, "rt", "aborted", 0)
+				scenario.requireJobsCount(t, "rt", "failed", 0)
+
+				scenario.requireTotalJobsCount(t, "rt", 0)
+
+				// reporting metrics
+				scenario.requireReportsFromBotManagement(t, botManagementReportExpectations{eventType: "track", eventName: "TestEvent", botDroppedCount: 1})
+				scenario.requireReportsFromBotManagement(t, botManagementReportExpectations{eventType: "identify", eventName: "", botDroppedCount: 1})
+				scenario.requireReportsFromBotManagement(t, botManagementReportExpectations{eventType: "screen", eventName: "", botDroppedCount: 1})
```

Comment:
One thing I see with these tests is that, it could be possible that reports could be flushed to Reporting Service before these validations

Reply:
Need not fix it as part of this pr, but there is one more issue with these metrics. These reporting metrics are sent to the production reporting instance. I think we should fix it.